### PR TITLE
Feature: Show the cargoes the vehicles can carry in the list window.

### PR DESCRIPTION
**Pull request by stormcone**
This patch is based on  @KeldorKatarn's commit:
KeldorKatarn/OpenTTD_PatchPack@0a57222a

cargo

### CODE REVIEW

**Comment by James103** [comment]
Should there be a setting to toggle the effects of this PR? Not all players may want this change as it may increase the clutter for some.

**Comment by TrueBrain** [comment]
A setting for this would be the right call, yes, but it can be on by default as far as I am concerned.

You did mangle two patches here, it seems, as I think part of "Feature: Show group name as part of the default vehicle name." is also in here :)

**Comment by TrueBrain** [comment]
> 
> You did mangle two patches here, it seems, as I think part of "Feature: Show group name as part of the default vehicle name." is also in here :)

I fixed my issue by merging your other patch. A rebase solves the problem now :D

File: src/vehicle_gui.cpp
Comment (TrueBrain) [inline src/vehicle_gui.cpp]: Are these comments correct? Or are they a copy/paste? :)

File: src/vehicle_gui.cpp
Comment (TrueBrain) [inline src/vehicle_gui.cpp]: variables can go inside the if-scope now I guess :)

File: src/vehicle_gui.cpp
Comment (stormcone) [reply src/vehicle_gui.cpp]: All of the 3 comments are different and I think they are correct. In the last one the first part of the string is filled with an empty string so it will only display the cargoes.

File: src/vehicle_gui.cpp
Comment (stormcone) [reply src/vehicle_gui.cpp]: Indeed. I tried whether I can put the string concatenation part into the inner branches, but it was starting to be messy and always crashed the game, so I gave up and stayed with this solution. :)

**Comment by frosch123** [comment]
Is there somewhere an explanation what this info is used for?
* Usually all vehicles in a group carry the same cargo, unless they use refit orders, right?
* Cloning vehicles copies refitting, so it's also unlikely to introduce inconsistencies?

I would see the usefulness of prepending groups names with the cargos of all member vehicles, but I don't understand why someone would want to see it per vehicle.

**Comment by stormcone** [comment]
Its depend on how someone organizes groups or whether using groups at all. The vehicles can be grouped by engines, by stations they visit, etc. But may be it is useful in the "all group" group. For example if you want to see which trains make the lowest profits, you can also whether the cargo makes them not profitable, so maybe better to carry it by other vehicle types.

But maybe sometimes you just would like to see which vehicle carry what when you just browsing the vehicle list. :)

I admit it may be the 'most' useful in the "all group" group. :)

Edit:
It also shows the cargoes in the station and waypoint lists:
openttd_cargo_names

**Comment by 2TallTyler** [comment]
This seems like a useful tool for waypoint and station management, especially with NewGRFs with lots of cargos carried in similar vehicles (looking at you, Iron Horse + FIRS Steeltown).

It needs some padding between the text and the vehicle sprite. Default engines abut the text with no space in between.

padding

Also needs a rebase (hopefully the last one, because I'd like to approve this 😃 )

**Comment by stormcone** [comment]
> It needs some padding between the text and the vehicle sprite. Default engines abut the text with no space in between.

Its using the same string as the vehicles' names and also depends on the sprites' offsets. (Give a name to a vehicle, and check it out it in the vehicle list.)

vehicle_name

So I would like to leave this to someone else with much more experience with padding, margins, and scaling. :)

**Comment by 2TallTyler** [comment]
@PeterN Any thoughts on the padding issue described above? Apparently this is an existing problem. Will #10114 fix it? 🙂 

(Once that's solved, I'd like to approve this.)

**Comment by Arastais** [comment]
Maybe out of the scope of this PR, but I think it'd be really nice if you could put the cargo list dynamically in the name with a string key. For example if you put `Train 1 for {CARGO}` in the name editor box for a vehicle it would show up everywhere else as "Train 1 for [Iron, Wood]". or maybe don't include the brackets and just make `Train 1 for ({CARGO})` show up as "Train 1 for (Iron, Wood)"

**Review by 2TallTyler — DISMISSED** [review]
My bad, the text abutting the vehicle sprite is a problem inherent to OpenGFX graphics with the default font. It happens outside this PR. (I use original TTD baseset which doesn't have this problem). That's not your problem to fix, and it looks like #10114 solves it anyway.

OpenGFX graphics and default font in current nightly:
nightly

Adding a custom cargo flag for player-entered strings is definitely out of scope. 🙂 

As I explained above, I think this is a useful feature for managing busy networks with lots of cargos. I'd use it. (And those who don't like the clutter can simply disable it.)

File: src/lang/english.txt
Comment (LordAro) [inline src/lang/english.txt]: Unicode apostrophe won't work so well with the sprite font

File: src/table/settings/gui_settings.ini
Comment (LordAro) [inline src/table/settings/gui_settings.ini]: Should this default to true, I wonder?

**Review by LordAro — DISMISSED** [review]

**Review by 2TallTyler — COMMENTED** [review]
When a vehicle doesn't have its own name or a group name, the cargos listed alone begin with a space. I wonder if the space can be appended by code instead of being hardcoded in the language file, so it can only be included if needed to separate name and cargos.

space

**Review by 2TallTyler — APPROVED** [review]

**Review by LordAro — APPROVED** [review]

**Pull request merged**
